lib/ddtrace/profiling/ext/forking.rb

```diff
@@ -37,20 +37,35 @@ module Kernel
           FORK_STAGES = [:prepare, :parent, :child].freeze
 
           def fork
-            wrapped_block = proc do
-              # Trigger :child callback
-              at_fork_blocks[:child].each(&:call) if at_fork_blocks.key?(:child)
-              yield
-            end
+            # If a block is provided, it must be wrapped to trigger callbacks.
+            child_block = if block_given?
+                            proc do
+                              # Trigger :child callback
+                              at_fork_blocks[:child].each(&:call) if at_fork_blocks.key?(:child)
+                              yield
+                            end
+                          end
 
             # Trigger :prepare callback
             at_fork_blocks[:prepare].each(&:call) if at_fork_blocks.key?(:prepare)
 
             # Start fork
-            result = super(&wrapped_block)
+            # If a block is provided, use the wrapped version.
+            result = child_block.nil? ? super : super(&child_block)
+
+            # Trigger correct callbacks depending on whether we're in the parent or child.
+            # If we're in the fork, result = nil: trigger child callbacks.
+            # If we're in the parent, result = fork PID: trigger parent callbacks.
+            # rubocop:disable Style/IfInsideElse
+            if result.nil?
+              # Trigger :child callback
+              at_fork_blocks[:child].each(&:call) if at_fork_blocks.key?(:child)
```

Comment:
Will `at_fork_blocks[:child].each(&:call)` trigger twice if a block is provided? (first when we wrap the block, then again here)

Reply:
@marcotc This is a great question but I think the answer is "no" (and please do check my logic here.)

When you call `fork` with a block, it will return the PID of the fork. When you call `fork` without a block, it will return in the child with `nil`, and with the PID of the fork in the parent.

If a block is passed, then `result` can never be `nil`. The wrapped block instead will run which will trigger `:child` callbacks, but not this statement.

If a block is not passed, then `result` may or may not be `nil`. But there won't be a wrapped block with `:child` callbacks, so in the parent the result will not be `nil` and this won't run, and in the child it will be `nil` and this statement will run.

Does that check out to you?

Reply:
It checks out. I was under the impression that, in the version of `fork` with a block provided, the child process would eventually run the block but then continue execution, which is not the case (after the block is done, the subprocess ends).

I believe the logic is sounds here then, thank you for the clarification!